File: DP/E-1137. N-th Tribonacci Number/E-1137. N-th Tribonacci Number.py
class Solution:
    @classmethod
    def tribonacci(cls, n: int) -> int:
        cache = {}
        return Solution.calc_tribonacci(n, cache)

    @classmethod
    def calc_tribonacci(cls, n: int, cache: dict):
        if n == 0:
            return 0
        if n == 1 or n == 2:
            return 1
        if n in cache:
            return cache.get(n)

        result = (
            Solution.calc_tribonacci(n - 1, cache)
            + Solution.calc_tribonacci(n - 2, cache)
            + Solution.calc_tribonacci(n - 3, cache)
        )
        cache[n] = result
        return result


testData = [
    {"n": 4, "ans": 4},
    {"n": 25, "ans": 1389537},
]

for data in testData:
    print(data["ans"] == Solution.tribonacci(data["n"]))

File: DP/E-1137. N-th Tribonacci Number/E-1137. N-th Tribonacci Number.ts
function tribonacci(n: number): number {
  const cache: Map<number, number> = new Map();
  return getTribonacci(n, cache);
}

const getTribonacci = (n: number, cache: Map<number, number>): number => {
  if (n === 0) return 0;
  if (n === 1 || n === 2) return 1;

  if (cache.has(n)) return cache.get(n) as number;
  const result =
    getTribonacci(n - 3, cache) +
    getTribonacci(n - 2, cache) +
    getTribonacci(n - 1, cache);
  cache.set(n, result);
  return result;
};

const testData = [
  { n: 4, ans: 4 },
  { n: 25, ans: 1389537 },
];

for (const data of testData) {
  console.log(data['ans'] === tribonacci(data['n']));
}
